gc.c

```diff
@@ -5180,16 +5183,27 @@ check_stack_for_moved(rb_objspace_t *objspace)
     each_machine_stack_value(ec, revert_machine_stack_references);
 }
 
-static void
-gc_compact_finish(rb_objspace_t *objspace, rb_size_pool_t *pool, rb_heap_t *heap)
+static bool
+gc_uncompacted_pools_p(rb_objspace_t *objspace)
 {
     for (int i = 0; i < SIZE_POOL_COUNT; i++) {
         rb_size_pool_t *size_pool = &size_pools[i];
-        rb_heap_t *heap = SIZE_POOL_EDEN_HEAP(size_pool);
-        gc_unprotect_pages(objspace, heap);
+        if (!size_pool->compacted) {
+            return TRUE;
+        }
     }
+    return FALSE;
+}
 
-    uninstall_handlers();
+static void
+gc_compact_finish_pool(rb_objspace_t *objspace, rb_size_pool_t *pool, rb_heap_t *heap)
```

Comment:
I think the current steps look like this:

1. Compact size pool 0
2. Finish compact size pool 0 (+ update references for all size pools)
3. Sweep size pool 0
4. Compact size pool 1
5. Finish compact size pool 1 (+ update references for all size pools)
6. Sweep size pool 1
...

What I think we can instead do is:

1. Compact size pool 0
2. Compact size pool 1
...
3. Finish compact all size pools
4. Sweep size pool 0
5. Sweep size pool 1
...

This would mean we only need to ever update references once. This would also make implementing cross size pool compaction easier in the future.